py/qstr.c

```diff
@@ -171,14 +194,48 @@ STATIC qstr qstr_add(const byte *q_ptr) {
     return MP_STATE_VM(last_pool)->total_prev_len + MP_STATE_VM(last_pool)->len - 1;
 }
 
+#define MP_QSTR_SEARCH_THRESHOLD 10
+
 qstr qstr_find_strn(const char *str, size_t str_len) {
-    // work out hash of str
     mp_uint_t str_hash = qstr_compute_hash((const byte *)str, str_len);
 
     // search pools for the data
     for (qstr_pool_t *pool = MP_STATE_VM(last_pool); pool != NULL; pool = pool->prev) {
-        for (const byte **q = pool->qstrs, **q_top = pool->qstrs + pool->len; q < q_top; q++) {
-            if (Q_GET_HASH(*q) == str_hash && Q_GET_LENGTH(*q) == str_len && memcmp(Q_GET_DATA(*q), str, str_len) == 0) {
+        size_t low = 0;
+        size_t high = pool->len - 1;
+
+        // binary search inside the pool
+        if (pool->sorted) {
+            while (high - low > MP_QSTR_SEARCH_THRESHOLD) {
+                size_t mid = (low + high + 1) / 2;
+                const byte **q = pool->qstrs + mid;
+                size_t len = Q_GET_LENGTH(*q);
+                if (len > str_len) {
+                    len = str_len;
+                }
+                int cmp = memcmp(Q_GET_DATA(*q), str, str_len);
```

Comment:
It might work to sort on qstr hash (and length) instead of qstr data, and then this comparison here would just be a comparison (subtraction) of the 2 hashes, which would be a lot faster than a memcmp.

Once it reached a case of 2 identical hashes it would then need to use memcmp, or just fall through to a linear search.

Reply:
This is a good idea so I gave it a shot.  

It turned out to be more challenging than I expected:
As you can see, [extra_coverage](https://github.com/micropython/micropython/runs/5325683711?check_suite_focus=true) test fails. That's because it tests auto completions which prints qstrs in order(`find_completions`, `print_completions`). So once qstrs are sorted by hash, the autocompletion order is no longer alphabetical...

We have a few options:
- Sort the auto completions results (allocate ram and sort qstr range by name?)
- Keep autocompletions unsorted
- Give up sorting by hash and keep alphabetical qstr sort

What do you think?

Reply:
Would be good to run some benchmarks on sorting by hash vs string content.  Do you have the original version which used string sorting?

Reply:
> Do you have the original version which used string sorting?

Yes. Only the last commit on this PR introduced hash sorting.  
Simply checkout previous commmit 5808635e4